gk/fib.c

```diff
@@ -63,40 +64,83 @@ static inline int
 initialize_ether_cache(struct ether_cache *eth_cache)
 {
 	memset(eth_cache, 0, sizeof(*eth_cache));
+	seqlock_init(&eth_cache->lock);
 
 	return 0;
 }
 
-/*
- * Fill up the Ethernet cached header.
- * Note that the destination MAC address should be filled up by LLS.
- */
-static inline void
+static void
+gk_arp_and_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	struct ether_cache *eth_cache = arg;
+
+	/*
+	 * Deal with concurrency control by sequential lock
+	 * on the nexthop entry.
+	 */
+	write_seqlock(&eth_cache->lock);
+
+	if (map->stale) {
+		ether_addr_copy(&map->ha, &eth_cache->eth_hdr.d_addr);
+		eth_cache->stale = false;
+	}
+
+	write_sequnlock(&eth_cache->lock);
+
+	if (pcall_again != NULL)
+		*pcall_again = true;
+}
+
+/* Fill up the Ethernet cached header. */
+static inline int
 fill_up_ether_cache_locked(struct ether_cache *eth_cache,
-	struct ipaddr *addr, struct gatekeeper_if *iface)
+	struct ipaddr *addr, struct gatekeeper_if *iface, int lcore_id)
 {
+	write_seqlock(&eth_cache->lock);
 	eth_cache->stale = true;
 	rte_memcpy(&eth_cache->ip_addr, addr, sizeof(eth_cache->ip_addr));
 	eth_cache->eth_hdr.ether_type = addr->proto;
 	ether_addr_copy(&iface->eth_addr, &eth_cache->eth_hdr.s_addr);
 	eth_cache->ref_cnt = 1;
+	write_sequnlock(&eth_cache->lock);
+
+	if (addr->proto == ETHER_TYPE_IPv4) {
+		return hold_arp(gk_arp_and_nd_req_cb,
+			eth_cache, &addr->ip.v4, lcore_id);
+	} else {
+		return hold_nd(gk_arp_and_nd_req_cb,
+			eth_cache, &addr->ip.v6, lcore_id);
+	}
 }
 
 static struct ether_cache *
 neigh_get_ether_cache_locked(struct neighbor_hash_table *neigh,
-	struct ipaddr *addr, struct gatekeeper_if *iface)
+	struct ipaddr *addr, struct gatekeeper_if *iface, int lcore_id)
 {
 	int ret;
 	struct ether_cache *eth_cache = lookup_ether_cache(neigh, &addr->ip);
 	if (eth_cache != NULL) {
+		write_seqlock(&eth_cache->lock);
 		eth_cache->ref_cnt++;
+		write_sequnlock(&eth_cache->lock);
+
 		return eth_cache;
 	}
 
 	ret = rte_hash_add_key(neigh->hash_table, &addr->ip);
 	if (ret >= 0) {
+		int succ;
 		eth_cache = &neigh->cache_tbl[ret];
-		fill_up_ether_cache_locked(eth_cache, addr, iface);
+		succ = fill_up_ether_cache_locked(
```

Comment:
I see -- there is an exclusive lock on the table, and then a lock on each individual entry because it might be being read by another thread. Got it, thanks!